drivers/rn2xx3/rn2xx3_internal.c

```diff
@@ -0,0 +1,317 @@
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+    _terminate_command(dev);
+
+    if (rn2xx3_wait_response(dev) == RN2XX3_TIMEOUT) {
+        mutex_unlock(&(dev->cmd_lock));
+        return RN2XX3_TIMEOUT;
+    }
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    rn2xx3_set_internal_state(dev, prev_state);
+
+    return rn2xx3_process_response(dev);
+}
+
+int rn2xx3_write_cmd_no_wait(rn2xx3_t *dev)
+{
+    DEBUG("[rn2xx3] CMD: %s\n", dev->cmd_buf);
+
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    mutex_lock(&(dev->cmd_lock));
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+    _terminate_command(dev);
+
+    DEBUG("[rn2xx3] RET: %s\n", dev->resp_buf);
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    return rn2xx3_process_response(dev);
+}
+
+int rn2xx3_wait_response(rn2xx3_t *dev)
+{
+    if (_wait_reply(dev, RESP_TIMEOUT_SEC)) {
+        return RN2XX3_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] RESP: %s\n", dev->resp_buf);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_wait_reply(rn2xx3_t *dev, uint8_t timeout)
+{
+    if (_wait_reply(dev, timeout)) {
+        return RN2XX3_REPLY_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] REPLY: %s\n", dev->resp_buf);
+
+    return rn2xx3_process_reply(dev);
+}
+
+void rn2xx3_cmd_start(rn2xx3_t *dev)
+{
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_CMD);
+    DEBUG("[rn2xx3] CMD: %s", dev->cmd_buf);
+    mutex_lock(&(dev->cmd_lock));
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+}
+
+void rn2xx3_cmd_append(rn2xx3_t *dev, uint8_t *payload, uint8_t payload_len)
+{
+    char payload_str[2];
+    rn2xx3_bytes_to_hex(payload, payload_str, payload_len);
+    DEBUG("%s", payload_str);
+    uart_write(dev->p.uart, (uint8_t*)payload_str, payload_len * 2);
+}
+
+int rn2xx3_cmd_finalize(rn2xx3_t *dev)
+{
+    DEBUG("\n");
+
+    uart_write(dev->p.uart, (uint8_t*)"\r\n", 2);
+    if (rn2xx3_wait_response(dev) == RN2XX3_TIMEOUT) {
+        rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_IDLE);
+        mutex_unlock(&(dev->cmd_lock));
+        return RN2XX3_TIMEOUT;
+    }
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    return RN2XX3_OK;
+}
+
+void rn2xx3_mac_tx_start(rn2xx3_t *dev)
+{
+    memset(dev->cmd_buf, 0, strlen(dev->cmd_buf));
+    snprintf(dev->cmd_buf, 7, "mac tx");
+    strcat(dev->cmd_buf, " ");
+    switch (dev->loramac.tx_mode) {
+        case LORAMAC_TX_CNF:
+            strcat(dev->cmd_buf, "cnf");
+            break;
+
+        case LORAMAC_TX_UNCNF:
+            strcat(dev->cmd_buf, "uncnf");
+            break;
+
+        default:
+            DEBUG("[rn2xx3] unsupported join procedure mode\n");
+            break;
+    }
+
+    strcat(dev->cmd_buf, " ");
+
+    char str_port[4];
+    snprintf(str_port, 3, "%d", dev->loramac.tx_port);
+
+    strcat(dev->cmd_buf, str_port);
+    strcat(dev->cmd_buf, " ");
+
```

Comment:
Now we got this far, how can we be sure cmd_buf did not overflow?

Reply:
Please add the same trick as you did for the other snprintf's

    dev->cmd_buf[p] = 0;

Reply:
This is already removed. It seems that you were looking at old stuff here.

Reply:
Hmm, not sure. Perhaps Github is confusing us.
I'm talking about the snprintf in rn2xx3_mac_tx_start. All other uses of snprintf are followed by writing a 0 in dev->cmd_buf[p], but not here.